From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: queenkjuul <queenkjuul@pm.me>
Date: Tue, 28 Jul 2026 19:12:43 +0200
Subject: [PATCH] cvitek: cv181x: enable distroboot

Adds BOOT_TARGET_DEVICES and the distro_bootcmd environment to the shared
cv181x board config, and makes CONFIG_BOOTCOMMAND try distro_bootcmd before
falling back to the vendor 'run sdboot'.

Without this the vendor U-Boot never looks for /extlinux/extlinux.conf, which
is what Armbian writes when SRC_EXTLINUX=yes.

booti() decompresses into kernel_comp_addr_r after pxe has already loaded the
FDT and the ramdisk, so that buffer has to sit above both. The vendor uImage
window does not: CVIMMAP_UIMAG_ADDR + CVIMMAP_UIMAG_SIZE is 0x81800000 plus
16MB, which covers fdt_addr_r, fdtoverlay_addr_r, pxefile_addr_r and
scriptaddr, so it is 0x88000000 here instead.

Originally from queenkjuul's milkv-bootloader patch series
(https://github.com/queenkjuul/milkv-duo-ubuntu), re-rooted onto the
sophgo/u-boot-2021.10 tree.
---
 include/configs/cv181x-asic.h | 22 +++++++++++++++++++---
 1 file changed, 19 insertions(+), 3 deletions(-)

diff --git a/include/configs/cv181x-asic.h b/include/configs/cv181x-asic.h
index 703aa745..12fae834 100644
--- a/include/configs/cv181x-asic.h
+++ b/include/configs/cv181x-asic.h
@@ -9,6 +9,7 @@
 #define __CV181X_ASIC_H__
 
 #include <../../../board/cvitek/cv181x/cv181x_reg.h>
+#include <config_distro_bootcmd.h>
 
 /* defined in cvipart.h */
 #undef CONFIG_ENV_OFFSET
@@ -187,6 +188,11 @@
 #define CONFIG_GATEWAYIP		192.168.0.11
 #define CONFIG_SERVERIP			192.168.56.101
 
+#define BOOT_TARGET_DEVICES(func) \
+	func(MMC, mmc, 0) \
+	func(DHCP, dhcp, 0) \
+	func(PXE, pxe, na)
+
 #ifdef CONFIG_USE_DEFAULT_ENV
 /* The following Settings are chip dependent */
 /******************************************************************************/
@@ -281,7 +287,17 @@
 		"root=" ROOTARGS "\0" \
 		"sdboot=" SD_BOOTM_COMMAND "\0" \
 		"othbootargs=" OTHERBOOTARGS "\0" \
-		PARTS_OFFSET
+		"kernel_addr_r=" __stringify(CONFIG_SYS_TEXT_BASE) "\0" \
+		"fdt_addr_r=0x82200000\0" \
+		"fdtoverlay_addr_r=0x82300000\0" \
+		"pxefile_addr_r=0x82400000\0" \
+		"scriptaddr=0x82500000\0" \
+		"ramdisk_addr_r=0x84000000\0" \
+		"kernel_comp_addr_r=0x88000000\0" \
+		"kernel_comp_size=0x1000000\0" \
+		"fdtfile=" FDT_NO ".dtb\0" \
+		PARTS_OFFSET \
+		BOOTENV
 
 /********************************************************************************/
 	/* UBOOT_VBOOT commands */
@@ -328,10 +344,10 @@
 		#ifdef CONFIG_ENABLE_ALIOS_UPDATE
 			#define CONFIG_BOOTCOMMAND	"cvi_update_rtos"
 		#else
-			#define CONFIG_BOOTCOMMAND	SHOWLOGOCMD "cvi_update || run norboot || run nandboot ||run emmcboot"
+			#define CONFIG_BOOTCOMMAND	"cvi_update || run distro_bootcmd || run norboot || run nandboot ||run emmcboot"
 		#endif
 	#else
-		#define CONFIG_BOOTCOMMAND	SHOWLOGOCMD "run sdboot"
+		#define CONFIG_BOOTCOMMAND	"run distro_bootcmd || run sdboot"
 	#endif
 
 	#if defined(CONFIG_NAND_SUPPORT)
